Barebones resource oriented API (#398)
ietf-wg-ppm/draft-ietf-ppm-dap: draft-ietf-ppm-dap.md

* Computing the results of a given measurement, specified in {{aggregate-flow}}
* Collecting aggregated results, specified in {{collect-flow}}

> In this section, we discuss each of the resources and the messages used to act
> on these resources. A resource's path is resolved relative to a server's
> endpoint to construct a resource URI.

Reviewer: Suggested rephrasing, to introduce the resources up front:
```suggestion
In this section, we discuss each of these interactions, the resources they act on (report, aggregation job, and collection job), and the messages used to do so. A resource's path is resolved relative to a server's
endpoint to construct a resource URI.
```

Author: I don't want to have to keep this list of resources in sync with the remainder of the document.

Resource paths are specified as templates like:

~~~
{role}/resource_type/{resource-id}
~~~

`{role}` is one of the API endpoints in the task's `aggregator_endpoints` (see
{{task-configuration}}). The remainder of the path is resolved relative to the
endpoint.

> DAP resource identifiers are always opaque byte strings, so anywhere
> `{resource-id}` occurs in a template (e.g., `{task-id}` or `{report-id}`) MUST
> be expanded to the URL-safe, unpadded Base 64 representation of the resource's
> identifier, as specified in sections 5 and 3.2 of {{!RFC4648}}.

Reviewer: rephrasing:
```suggestion
DAP resource identifiers are opaque byte strings, so any occurrence of
`{resource-id}` in a URL template (e.g., `{task-id}` or `{report-id}`) MUST
be expanded to the URL-safe, unpadded Base 64 representation of the corresponding resource
identifier, as specified in sections 5 and 3.2 of {{!RFC4648}}.
```

* Computing the results of a given measurement, specified in {{aggregate-flow}}
* Collecting aggregated results, specified in {{collect-flow}}

In this section, we discuss each of the resources and the messages used to act
on these resources. A resource's path is resolved relative to a server's

> endpoint to construct a resource URI. A server's endpoint MAY contain
> arbitrarily many path components before the DAP resources.

Reviewer 2: This guidance doesn't seem necessary to me given the definition of `[role]` below.

```suggestion
endpoint to construct a resource URI.
```

Resource paths are specified as templates like:

~~~

> [role]/resource_type/{resource-id}

Reviewer 2: Suggestion: For consistency with the other "variables" in paths, use curly braces instead of square braces. Unless, do you intend `{}` to denote a variable that is expanded into base64 encoding?
```suggestion
{role}/resource_type/{resource-id}
```

Author: No, I was just preserving the existing usage of `[leader]`, `[helper]` and `[aggregator]`. Using `{}` is more consistent, though.

Reviewer 2: Consistency is a good idea, I think. The curly brackets look nicer.

### Upload Request

Clients upload reports by using an HTTP POST to `[leader]/upload`, where
`[leader]` is the first entry in the task's aggregator endpoints. The payload is
structured as follows:
Clients upload reports by using an HTTP PUT to
`[leader]/tasks/{task-id}/reports/{report-id}`, where `[leader]` is the first
entry in the task's aggregator endpoints.

`task-id` is the ID of the task to which the report is being uploaded.
`report-id` is a `ReportId` used by the Aggregators to ensure the report appears
in at most one batch (see {{input-share-validation}}). The Client MUST generate
this using a cryptographically secure random number generator.

[Note/TODO: it's important that the _Client_ chooses the report ID. We use the
report ID as the nonce, and VDAF-04 will include some guidance explaining that
if the report ID is chosen with knowledge of the VDAF verify key, privacy could
be violated. For that reason, we can't allow either aggregator to choose report
IDs.]

The payload is structured as follows:

~~~
struct {

>   ReportID report_id;

Reviewer 2: Hrmm, I'm a bit wary. of moving the report ID out of the `Report` struct. Can you explain your thinking here?

In any case, we need to make sure that the task ID, report ID, timestamp, and any other metadata that we might have in the future are all included in the report metadata.

Author: The report ID now occurs in the request path `[leader]/tasks/{task-id}/reports/{report-id}`, so it does not need to occur in the request body. `struct ReportMetadata` does still contain `ReportID` so that it can be used to construct `InputShareAad` and `ReportShare`.

I suppose it's not essential that the report ID be in the request path: thus far, it's only the task ID that's needed by servers to figure out how to parse the rest of the message. So we could put report ID back into the body, and then we can use `ReportMetadata` in `Report` again, which is a little nicer.

Author: The downside of this change is that it's inconsistent with the URIs for aggregation jobs and collections, but I don't really think that matters.

Reviewer 2: My vote would be to move report ID back into body so that the metadata appears in both report and report share. If you decide to to keep the report ID where it is, then my only request is to make sure to spell out how the metadata is constructed where we specify encryption of the input shares.

Reviewer 2: Oh looks like you already updated this! Thanks.

  opaque public_share<0..2^32-1>;
  HpkeCiphertext encrypted_input_shares<1..2^32-1>;
} Report;
~~~

* `report_metadata` is public metadata describing the report.
    * `report_id` is used by the Aggregators to ensure the report appears in at
      most one batch (see {{input-share-validation}}). The Client MUST generate
      this by generating 16 random bytes using a cryptographically secure random
      number generator.

[Note/TODO: it's important that the _Client_ chooses the report ID. We use the
report ID as the nonce, and VDAF-04 will include some guidance explaining that
if the report ID is chosen with knowledge of the VDAF verify key, privacy could
be violated. For that reason, we can't allow either aggregator to choose report
IDs.]

    * `time` is the time at which the report was generated. The Client SHOULD
      round this value down to the nearest multiple of the task's
      `time_precision` in order to ensure that that the timestamp cannot be used
      to link a report back to the Client that generated it.

* `public_share` is the public share output by the VDAF sharding algorithm. Note
  that the public share might be empty, depending on the VDAF. For example,

>   Prio3 has an empty public share, but Poplar1 does not. See {{!VDAF}}.

Author: We can make that change in a separate PR where DAP will adapt to VDAF-04, but I want to keep this PR focused on its stated goal.

query and be ready to return the aggregate shares right away, but this cannot be
guaranteed. In fact, for some VDAFs, it is not be possible to begin preparing
inputs until the collector provides the aggregation parameter in the
`CollectionReq`. For these reasons, collect requests are handled asynchronously.

Upon receipt of a `CollectionReq`, the leader begins by checking that it
recognizes the task ID in the request path. If not, it MUST abort with error
`unrecognizedTask`. Then, the leader verifies that the request meets the
requirements of the batch parameters using the procedure in
{{batch-validation}}. If so, it immediately responds with HTTP status 201.

The leader then begins working with the helper to prepare the shares satisfying
the query (or continues this process, depending on the VDAF) as described in
{{aggregate-flow}}.

After receiving the response to its `CollectionReq`, the collector makes an HTTP
`POST` request to the collection job URI to check on the status of the collect
job and eventually obtain the result. If the collection  job is not finished
yet, the leader responds with HTTP status 202 Accepted. The response MAY include

> a Retry-After header field to suggest a pulling interval to the collector.

Reviewer: pre-existing typo
```suggestion
a Retry-After header field to suggest a polling interval to the collector.
```

The message contains the following parameters:

* The DAP task ID. If the Helper does not recognize the task ID, it MUST abort

>   with error `unrecognizedTask`.

Reviewer 2: Reviewer note: This MUST has been moved below.
